torch/csrc/distributed/autograd/context/context.cpp

```diff
@@ -193,6 +202,28 @@ std::shared_ptr<rpc::FutureMessage> DistAutogradContext::
   return state->future;
 }
 
+void DistAutogradContext::setAutogradError(const std::string& errorMsg) {
+  setGraphTaskException(errorMsg);
+  propagateAutogradError(errorMsg);
+}
+
+void DistAutogradContext::setGraphTaskException(const std::string& errorMsg) {
+  auto graphTask = retrieveGraphTaskIdempotent();
+  if (graphTask) {
+    std::runtime_error err(errorMsg);
+    graphTask->set_exception(err, nullptr);
+  }
```

Comment:
This should move to request_callback_impl.cpp. Also, note that you are not holding the lock here.

Reply:
I kept this helper function here for now (updated to hold the lock like we do in `addOutstandingRpc` when setting an exception on `graphTask`). Should we still move it to `request_callback_impl.cpp` and set the exception outside of the lock?